initialize the distributions on the GPU

atmyers opened the pull request:
Parallelize particle initialization using `ParallelForRNG` for all backends.

Follow-up to #73
Close #495

- [x] ensure GPU results are correct

atmyers committed:
initialize the distributions on the GPU

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -0,0 +1,92 @@
+
+#include <type_traits>
+
+namespace impactx
+{
+namespace detail
+{
```
To match the directory structure:
```suggestion
namespace impactx::initialization
{
```

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -0,0 +1,92 @@
+        amrex::ParticleReal* const AMREX_RESTRICT m_part_py;
+        amrex::ParticleReal* const AMREX_RESTRICT m_part_pt;
+    };
+}
+}
```
```suggestion
}  // namespace impactx::initialization
```

ax3l commented on src/initialization/InitDistribution.cpp:
```diff
@@ -75,26 +74,27 @@ namespace impactx
+            auto py_ptr = py.data();
+            auto pt_ptr = pt.data();
+
+            auto init_single_particle_data = detail::InitSingleParticleData(distribution,
```
```suggestion
            auto init_single_particle_data = initialization::InitSingleParticleData(distribution,
```

ax3l committed:
Fix Thermal: `const` operator

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -0,0 +1,96 @@
+#include <AMReX_Extension.H> // for AMREX_RESTRICT
+#include <AMReX_REAL.H>
+
+#include <type_traits>
```
For `std::move`:
```suggestion
#include <utility>

```
`type_traits` is not used here.

ax3l committed:
Python Support, Cleaning

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -167,6 +167,83 @@ namespace impactx
+    {
+        BL_PROFILE("ImpactX::AddNParticles");
+
+        AMREX_ALWAYS_ASSERT_WITH_MESSAGE(lev == 0, "AddNParticles: only lev=0 is supported yet.");
```
ImpactX support MR now. I wonder if `lev!=0` is generally needed or if we want to remove the parameter altogether.

I assume that on `Redistribute` we automatically move the particles in the corresponding level at the end of `AddNParticles`?

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -167,6 +167,83 @@ namespace impactx
+            w_arr[old_np+i]  = bchchg/ablastr::constant::SI::q_e/np;
+        });
+
+        Redistribute();
```
At the end of the CPU `ImpactX::add_particles` we have:
```C++
       // Resize the mesh to fit the spatial extent of the beam and then
        // redistribute particles, so they reside on the MPI rank that is
        // responsible for the respective spatial particle position.
        this->ResizeMesh();
        m_particle_container->Redistribute();
```

I think the following bugs are here:
- If a user calls the device `AddNParticles`, it will forget to call `this->ResizeMesh();` before Redistribute. This will likely lead to particle loss because the grid extents are not up to date.
- If a user calls the host `AddNParticles`, then `Redistribute` will becalled twice.

ax3l committed:
Cleanup: Unused, Resize

Fix the missing resize before redistribute.
Remove the old, unusued, serial initialization logic.

ax3l commented on src/initialization/InitDistribution.H:
```diff
@@ -0,0 +1,103 @@
+         * @param part_pt the array to the particle momentum (t)
+         */
+        InitSingleParticleData (
+            T_Distribution distribution,
```
Not a reference, since we want to copy because this object goes on the device.

ax3l committed:
Thermal Distribution: GPU Memory

Remove stack memory usage and use GPU memory for pre-computed CDF.

ax3l commented:
Ok, this passed the thermal tests on GPU on Perlmutter :tada:

ax3l commented:
@cemitch99 a couple of tests fail with this PR on GPU due to tolerances in analysis scripts. The main reason is that we change the way we draw random numbers and have different seeds in parallel execution now.

Unless there is a GPU bug, might need to relax tolerances a bit, I think.

ax3l commented:
```
The following tests FAILED:
	  2 - FODO.analysis (Failed)
	  5 - FODO.MPI.analysis (Failed)
	  8 - FODO.py.analysis (Failed)
	 11 - FODO.MADX.py.analysis (Failed)
	 14 - FODO.py.MPI.analysis (Failed)
	 17 - chicane.analysis (Failed)
	 20 - chicane.py.analysis (Failed)
	 23 - chicane.MADX.py.analysis (Failed)
	 34 - gaussian.analysis (Failed)
	 38 - FODO_RF.analysis (Failed)
	 40 - FODO_RF.py.analysis (Failed)
	 42 - kurth4d.analysis (Failed)
	 44 - semigaussian.analysis (Failed)
	 46 - multipole.analysis (Failed)
	 48 - expanding_beam.analysis (Failed)
	 50 - expanding_beam.py.analysis (Failed)
	 52 - multipole.py.analysis (Failed)
	 58 - iotalattice.MPI.analysis (Failed)
	 60 - iotalattice.py.MPI.analysis (Failed)
	 62 - kurth_periodic.analysis (Failed)
	 64 - kurth_periodic.py.analysis (Failed)
	 74 - solenoid.analysis (Failed)
	 76 - solenoid.py.analysis (Failed)
	 78 - solenoid.MADX.py.analysis (Failed)
	 88 - quadrupole_softedge.analysis (Failed)
	 90 - quadrupole_softedge.py.analysis (Failed)
	 92 - fodo_chromatic.analysis (Failed)
	 94 - fodo_chromatic.py.analysis (Failed)
	102 - cyclotron.py.analysis (Failed)
	110 - compression.py.analysis (Failed)
	136 - IOTA_lattice.analysis (Failed)
	138 - IOTA_lattice.py.analysis (Failed)
	143 - aperture.run (Failed)
	144 - aperture.analysis (Failed)
	145 - aperture.py.run (Failed)
	146 - aperture.py.analysis (Failed)
	148 - apochromat.analysis (Failed)
	152 - alignment.analysis (Failed)
```

ax3l commented:
For the FODO cell example, the initial beam sigma_x,y and emittance_x,y is 2 orders of magnitude smaller then the reference value.
That makes no real sense to me, since we set finite numbers. Could be a GPU bug.

Update: misread, indeed just sampling noise. Will slightly push tolerances.

ax3l commented on src/particles/ImpactXParticleContainer.cpp:
```diff
@@ -123,48 +125,55 @@ namespace impactx
+            pt_arr[old_np+i] = pt_ptr[i];
+            qm_arr[old_np+i] = qm;
+            w_arr[old_np+i]  = bchchg/ablastr::constant::SI::q_e/np;
+        });
```
Maybe needed now, although I would expect that `this->ResizeMesh();` should go in the right stream...
```suggestion
        });

        amrex::Gpu::streamSynchronize();
```

ax3l replied:
No change.

ax3l replied:
added for safety nonetheless (see inline comment)

ax3l committed:
Cosmetic Cleaning

ax3l committed:
Waterbag: Clean C++ Math

Avoid promoting to double in C functions.

ax3l committed:
Cosmetic Cleaning for Clarity

ax3l committed:
Fix CUDA Compile

On older NVCC

ax3l committed:
Analysis: Slightly Increase Sampling Tolerance

Run on local GPU (RTX A2000).
